src/core/execute.c

```diff
@@ -3294,7 +3302,19 @@ static int exec_child(
                  * wins here. (See above.) */
 
                 if (context->pam_name && username) {
+                        _cleanup_free_ gid_t *gids_before_pam;
+                        _cleanup_free_ gid_t *gids_after_pam;
+
+                        size_t ngids_before_pam = get_current_gids(&gids_before_pam);
                         r = setup_pam(context->pam_name, username, uid, gid, context->tty_path, &accum_env, fds, n_fds);
+                        size_t ngids_after_pam = get_current_gids(&gids_after_pam);
+
+                        // prepare for the worst case - all gids after pam has been added
+                        gids_added_by_pam = new(gid_t, ngids_after_pam);
+
+                        ngids_added_by_pam = find_gid_lists_diff(gids_before_pam, ngids_before_pam,
+                                gids_after_pam, ngids_after_pam, gids_added_by_pam);
```

Comment:
hmm, I don't think we really need to do the diffing. Looking at this again I think we should simplify things a bit. Let's do it like this:

1. If neither PAM nor SupplementaryGroups= are configured for a service, inherit group list from PID 1, i.e. do nothing
2. If SG= is set, but PAM is not, set the group list to the specified list in SG=, and don't inherit anything
3. If PAM is set, but SG= is not, then before invoking pam_setcreds set the group list to empty (only if we can, in containers we might not, which we should handle gracefully, and simple log at debug level about and continue), so that pam_setcreds has full effect and nothing is inherited from PID 1
4. If both PAM and SG= is configured, then reset the list before pam_secreds is set, and then merge the result with the list configured in SG=.

Or in other words: If neither is configured inherit the group list, if either is set, make that count, and disable inheriting. Only skewed slightly in containers where dropping gids is not allowed....

This should make things relatively simple I think, and something we can reasonably explain and document

Reply:
> 3. If PAM is set, but SG= is not, then before invoking pam_setcreds set the group list to empty (only if we can, in containers we might not, which we should handle gracefully, and simple log at debug level about and continue), so that pam_setcreds has full effect and nothing is inherited from PID 1

I'm not sure I'm getting that part correctly. What do you mean by "PAM is set"? E.g. with pam_groups you know that it altered the gids only from comparing groups before and after calling pam_setcreds. So setting group list to empty is not an option. Unless I'm missing something.

Reply:
by "PAM is set" I meant "PAMName=" being configured.

My suggestion is that before entering pam_setcreds() we'd reset the process credential's list of supplementary groups to empty, so that we know as a gurantee that what is set afterwards has been set by pam_setcreds().

Reply:
If I understand it correctly this would make it work only if PAMName is defined, so e.g. to fix [this bug](https://gitlab.gnome.org/GNOME/gdm/issues/393#) PAMName would have to be added to desktop entries of all apps started by DBus under user session.

I was aiming at a more generic solution that could guarantee correct pam_groups application of all units after adding pam_group once for all users/units e.g. at /etc/pam.d/.

Reply:
no, systemd --user services do not get to user User= or PAMName= at all. However, the systemd --user instance itself uses both, and that's where it matters.

Reply:
Oh, good. Thanks for clarifying that.

Reply:
I have pushed code simplified code. However, I'm still heaving issues with `systemd --user` sessions. Turns out user@UID.service hits this part of the code with PAMName defined as `systemd-user` and it gets the groups cleared e.g. for gnome-terminal DBus trigger launched as its child.
This happens in Fedora and Ubuntu. I'm not sure if I understood (and applied) your hints correctly.